File: src/Item29/java/Stack.java
package Item29.java;

import java.util.Arrays;
import java.util.EmptyStackException;

public class Stack<E> {
    private E[] elements;
    private int size = 0;
    private static final int DEFAULT_CAPACITY_SIZE = 16;

    @SuppressWarnings("unchecked")
    public Stack() {
        elements = (E[]) new Object[DEFAULT_CAPACITY_SIZE];
    }

    public void push(E e) {
        ensureCapacity();
        elements[size++] = e;
    }

    public E pop() {
        if (size == 0)
            throw new EmptyStackException();
        E result = elements[--size];
        elements[size] = null;

        return result;
    }

    private void ensureCapacity() {
        if (elements.length == size)
            elements = Arrays.copyOf(elements, 2 * size + 1);
    }

}

File: src/Item29/kotlin/Stack.kt
package Item29.kotlin

import java.util.*

class Stack<T : Any>(initialCapacity: Int = 16) {
    @Suppress("UNCHECKED_CAST")
    private var elements = arrayOfNulls<Any?>(initialCapacity) as Array<T?>
    private var size = 0

    val isEmpty = size == 0

    fun push(e: T) {
        ensureCapacity()
        elements[size++] = e
    }

    fun pop(): T {
        if (size == 0) throw EmptyStackException()
        return (elements[--size] as T).also {
            elements[size] = null // Eliminate obsolete reference
        }
    }

    private fun ensureCapacity() {
        if (elements.size == size)
            elements = Arrays.copyOf(elements, 2 * size + 1)
    }
}